File: java-core/src/main/java/com/minersecho/core/World.java
package com.minersecho.core;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

public final class World {
    private final Map<Vec3i, BlockType> blocks = new HashMap<>();
    private final BlockType defaultBlock;

    public World(BlockType defaultBlock) {
        this.defaultBlock = defaultBlock;
    }

    public BlockType getBlock(Vec3i position) {
        return blocks.getOrDefault(position, defaultBlock);
    }

    public void setBlock(Vec3i position, BlockType blockType) {
        blocks.put(position, blockType);
    }

    public boolean isAir(Vec3i position) {
        return getBlock(position) == BlockType.AIR;
    }

    public boolean isSolid(Vec3i position) {
        return !isAir(position);
    }

    public void fillAir(List<Vec3i> positions) {
        for (Vec3i position : positions) {
            setBlock(position, BlockType.AIR);
        }
    }
}

File: prototype/world.py
from dataclasses import dataclass, field


AIR = "air"
STONE = "stone"


@dataclass
class World:
    blocks: dict[tuple[int, int, int], str] = field(default_factory=dict)
    default_block: str = STONE

    def get_block(self, position: tuple[int, int, int]) -> str:
        return self.blocks.get(position, self.default_block)

    def set_block(self, position: tuple[int, int, int], block_type: str) -> None:
        self.blocks[position] = block_type

    def is_air(self, position: tuple[int, int, int]) -> bool:
        return self.get_block(position) == AIR

    def is_solid(self, position: tuple[int, int, int]) -> bool:
        return not self.is_air(position)

    def fill_air(self, positions: list[tuple[int, int, int]]) -> None:
        for position in positions:
            self.set_block(position, AIR)

    def fill_stone(self, positions: list[tuple[int, int, int]]) -> None:
        for position in positions:
            self.set_block(position, STONE)